software/contrib/conway.py

```diff
@@ -0,0 +1,354 @@
+        """
+        base_spawn_level = k1.percent() / 2
+
+        # get the level of AIN, attenuverted by K2
+        cv_mod = ain.percent()/2
+        cv_att = k2.percent() - 1
+        cv_mod = cv_mod * cv_att
+
+        spawn_level = clamp(base_spawn_level + cv_mod, 0, 1)
+        return spawn_level
+
+    def reset(self):
+        """Clear the whole field and spawn random data in it
+        """
+        for i in range(len(self.field)):
+            self.field[i] = 0x00
+            self.next_field[i] = 0x00
+
+        self.num_alive = 0
+        self.population_deltas = []
+
+        # fill the field with random cells
+        fill_level = self.calculate_spawn_level()
+        for i in range(NUM_PIXELS):
+            x = rnd()
+            is_alive = get_bit(self.field, i)
+            if x < fill_level and not is_alive:
+                # if the space isn't already filled and we want to fill it
+                set_bit(self.field, i, True)
+                set_bit(self.next_field, i, True)
+                self.num_alive = self.num_alive + 1
+            elif x >= fill_level and is_alive:
+                # if the space is filled and we want to clear it
+                set_bit(self.field, i, False)
+                set_bit(self.next_field, i, False)
+                self.num_alive = self.num_alive - 1
+
+        # Assume the whole field has changed
+        set_all_bits(self.changed_spaces, True)
+        self.num_changes = NUM_PIXELS
+
+    def draw(self):
+        """Show the current playing field on the OLED
+        """
+        oled.blit(self.frame, 0, 0)
+        oled.show()
+
+    def tick(self):
+        """Calculate the state of the next generation
+
+        This checks the regions around every changed space in the previous generation, updating the
+        total population and counting how many births & deaths this generation had.
+
+        If a reset was requested, the field is cleared & randomly reset _before_ calculating the new generation
+        """
+        self.num_born = 0
+        self.num_died = 0
+        self.num_changes = 0
+
+        if self.reset_requested:
+            self.reset_requested = False
+            self.reset()
+
+        for bit_index in range(NUM_PIXELS):
+            if get_bit(self.changed_spaces, bit_index):
+                neighbourhood = self.get_neigbour_indices(bit_index)
+                num_neighbours = 0
+                for n in neighbourhood:
+                    if get_bit(self.field, n):
+                        num_neighbours = num_neighbours + 1
+
+                if get_bit(self.field, bit_index):
+                    if num_neighbours == 2 or num_neighbours == 3:        # happy cell, stays alive
+                        set_bit(self.next_field, bit_index, True)
+                    else:                                                 # sad cell, dies
+                        set_bit(self.next_field, bit_index, False)
+                        self.num_died = self.num_died + 1
```

Comment:
[optional] Not a fan of `+=`? I feel like its use would slim down this code a bit, and make it slightly easier to read.

Reply:
I learned Python _before_ the `+=` and `-=` operators were added to it, so I just got in the (bad) habit of not using them in Python.